std/experimental/typecons.d

```diff
@@ -802,3 +802,223 @@ private template findCovariantFunction(alias finfo, Source, Fs...)
     else
         enum ptrdiff_t findCovariantFunction = x;
 }
+
+/++
+ + Type constructor for head-const variables.
+ +
+ + Head-const variables cannot be directly mutated or rebound, but references
+ + reached through the variable are typed with their original mutability.
+ + It is equivalent to $(D final) variables in D1 and Java, as well as
+ + $(D readonly) variables in C#.
+ +
+ + When $(D T) is a $(D const) or $(D immutable) type, $(D HeadConst) aliases
+ + to $(D T).
+ +/
+struct HeadConst(T)
+    if (!is(T == const) && !is(T == immutable))
+{
+    import std.typecons : Proxy;
+
+    private T HeadConst_value;
+    mixin Proxy!HeadConst_value;
+
+    /++
+     + Construction is forwarded to the underlying type.
+     +/
+    this(T other)
+    {
+        this.HeadConst_value = other;
+    }
+
+    /// Ditto
+    this(Args...)(auto ref Args args)
+        if (__traits(compiles, T(args)))
+    {
+        static assert((!is(T == struct) && !is(T == union)) || !isNested!T,
+            "Non-static nested type " ~ fullyQualifiedName!T ~ " must be " ~
+            "constructed explicitly at the call-site (e.g. auto s = " ~
+            "headConst(" ~ T.stringof ~ "(...));)");
+        this.HeadConst_value = T(args);
+    }
+
+    // Attaching function attributes gives less noisy error messages
```

Comment:
This actually sounds like something we should fix in the compiler. Is there a bugzilla entry for this?